pyccel/parser/syntax/openmp.py

```diff
@@ -11,396 +11,962 @@
+                raise TypeError(msg)
+            txt = '{0} {1}'.format(txt, clause.expr)
+        else:
+            msg = "Wrong clause " + type(clause).__name__
+            raise TypeError(msg)
+    return txt
+
+
 class Openmp(object):
     """Class for Openmp syntax."""
     def __init__(self, **kwargs):
-        """
-        Constructor for Openmp.
-
-        """
         self.statements = kwargs.pop('statements', [])
 
 class OpenmpStmt(BasicStmt):
-    """Class representing a ."""
+    """Class representing an OpenMP statement."""
     def __init__(self, **kwargs):
-        """
-        """
         self.stmt = kwargs.pop('stmt')
 
-        super(OpenmpStmt, self).__init__(**kwargs)
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
-            print("> OpenmpStmt: expr")
+            print("> {}: expr".format(type(self).__name__))
 
         stmt = self.stmt
-        if isinstance(stmt, OmpEndClause):
-            return stmt.expr
-        elif isinstance(stmt, OmpParallelConstruct):
-            return stmt.expr
-        elif isinstance(stmt, OmpLoopConstruct):
-            return stmt.expr
-        elif isinstance(stmt, OmpSingleConstruct):
+        if isinstance(stmt, omp_directives):
             return stmt.expr
         else:
             raise TypeError('Wrong stmt for OpenmpStmt')
 
-class OmpParallelConstruct(BasicStmt):
-    """Class representing a ."""
+class OmpParallelConstruct(OmpConstruct):
+    """Class representing a parallel construct."""
     def __init__(self, **kwargs):
-        """
-        """
-        self.clauses = kwargs.pop('clauses')
+        super().__init__(OMP_Parallel_Construct, _valid_parallel_clauses, **kwargs)
 
-        super(OmpParallelConstruct, self).__init__(**kwargs)
+class OmpLoopConstruct(OmpConstruct):
+    """Class representing a For loop construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_For_Loop, _valid_loop_clauses, **kwargs)
 
-    @property
-    def expr(self):
-        if DEBUG:
-            print("> OmpParallelConstruct: expr")
+class OmpTaskLoopConstruct(OmpConstruct):
+    """Class representing a taskloop construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, (_valid_taskloop_clauses + (OmpinReduction,)), **kwargs)
+
+class OmpTaskConstruct(OmpConstruct):
+    """Class representing a Task Construct """
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Task_Construct, _valid_task_clauses, **kwargs)
 
-        _valid_clauses = (OmpNumThread, \
-                         OmpDefault, \
-                         OmpPrivate, \
-                         OmpShared, \
-                         OmpFirstPrivate, \
-                         OmpCopyin, \
-                         OmpReduction, \
-                         OmpProcBind)
+class OmpSingleConstruct(OmpConstruct):
+    """Class representing a single construct."""
```

Comment:
typo: single -> Single, i think you should delete the letter 'a' in many cases because it indicates the literal meaning of the construct name and it is obvious in this case,

Reply:
I disagree. `Class representing Single construct.` is not correct English.

Maybe `a` should be changed to `the` though, `a` implies "one of several" I don't think there are multiple types of single constructs?

Reply:
actually, i removed 'replace it with `the`', but i considered `Single` as a proper name.

Reply:
If you count `Single` as a proper name then it would have to be:
`Class representing Single`
If you keep "construct" then it needs an article, so:
`Class representing the Single construct`
I find the second option clearer